Feature to add and delete multilingual text

lickem22 opened the pull request:
Reviewer: @sidravi1 
Estimate: 1h30

---

## Ticket

Fixes: [AAQ-450](https://idinsight.atlassian.net/browse/AAQ-450)

## Description

### Goal
The goal of this PR is to add the last feature of multilingual support which are adding a new content and deleting multilingual contents. 
### Changes
The following changes have made on the backend:
- Updated the` create endpoint `to only add new contents with the create content endpoint. Language versions will be added using the edit endpoint
- Updated the `edit endpoint` to create Language versions of contents.
- Updated `delete endpoint` to delete language version when a language_id is provided and delete the whole content otherwise. 
The following changes have been made on the backend:
- Added a new content using the New button
- Added a `delete button` on the` edit page` and the `Content Modal` to delete language versions of contents. If the last language version is deleted, the user is redirected to the content  landing page.
- Updated the back arrow to redirect to the content landing page. 
- Updated the add page to redirect to edit instead of the landing page when a new content is added/edited. 
### Future Tasks (optional)
With this PR, all features are implemented. The future tasks will be:
- Add backend tests
- Clean up (if necessary)
## How has this been tested?
This is the last multilingual feature PR, so all multilingual features have been tested:
- Adding a Content
- Adding a language version of the content (both from landing and Modal)
- Viewing different language versions of contents 
- Updating a content 
- Deleting a content 
- Deleting a Language version of a content
I suggest testing all this extensively while reviewing to make sure there is no additional changes to make to the backend. 

## Checklist

Fill with `x` for completed. Delete any lines that are not relevant

- [x] My code follows the style guidelines of this project
- [x] I have reviewed my own code to ensure good quality
- [x] I have tested the functionality of my code to ensure it works as intended
- [x] I have resolved merge conflicts
- [ ] I have updated the automated tests (if applicable)
- [ ] I have updated the requirements (if applicable)
- [ ] I have updated the README file (if applicable)
- [ ] I have updated affected documentation (if applicable)


[AAQ-450]: https://idinsight.atlassian.net/browse/AAQ-450?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

lickem22 commented on admin_app/src/components/LanguageButtonBar.tsx:
```diff
@@ -1,11 +1,9 @@
   appStyles,
   sizes,
 } from "@/utils";
-import { apiCalls } from "@/utils/api";
```
Removed apiCalls from the LanguageButtonBar as well

sidravi1 reviewed (changes requested):
Thanks @lickem22 - we are almost there. Left a few comments.

Also, can you please fix these errors in the console:
<img width="651" alt="image" src="https://github.com/IDinsight/aaq-core/assets/14125957/949b744d-ed78-43b5-b854-7da6e7ecb8f0">

<img width="901" alt="image" src="https://github.com/IDinsight/aaq-core/assets/14125957/b8cab8c6-920a-4f2c-a53c-af678ae3999d">

<img width="884" alt="image" src="https://github.com/IDinsight/aaq-core/assets/14125957/42ac23f9-9aa7-4b00-b144-906a61e89869">

lickem22 commented on admin_app/src/app/content/edit/page.tsx:
```diff
@@ -214,14 +294,32 @@ const ContentBox = ({
+        justifyContent="space-between"
+      >
+        <Typography variant="body2">Title</Typography>
+        <IconButton
```
Did that. But it still appears in edit mode when a language version is added is not added to the DB yet. I did that so there is no confusion in case someone picks a language by mistake and then click on another without saving the previous language version

lickem22 commented on admin_app/src/app/content/edit/page.tsx:
```diff
@@ -30,17 +31,36 @@ const AddEditContentPage = () => {
+    language_id: number | null,
+  ): string | null => {
+    if (action === "delete") {
+      return `Content #${content_id} with language_id:#${language_id} deleted successfully`;
```
So far, the content page doesn't communicate with language data. It would be easier to implement when we save languages to local storage. Would it be okay to add that as a separate ticket as well?

suzinyou replied:
Let's do that. Could you create a ticket for this and link to Sid's comment? Thanks!!

lickem22 commented on admin_app/src/app/content/edit/page.tsx:
```diff
@@ -138,6 +199,11 @@ const ContentBox = ({
       language_id: content.language_id,
       content_metadata: content.content_metadata,
     };
+    if (body.language_id === 0) {
```
Updated the add page to use the default language!

suzinyou replied:
Yes, we should store the languages locally. Let's link this comment also in the ticket!

suzinyou replied:
But we should address it before any serious (external!)deployment happens..!

lickem22 replied:
I don't mind doing it next sprint. As long as it is not as part of this feature.

lickem22 commented on admin_app/src/app/content/edit/page.tsx:
```diff
@@ -138,6 +199,11 @@ const ContentBox = ({
       content_metadata: content.content_metadata,
     };
+    if (body.language_id === 0) {
+      setErrorText("Please select a language");
```
I am assuming you think it shouldn't do that because if you are adding a new language and then noticed it is the wrong language, everything will deleted. But I did that   because I noticed that if I don't, when you go from an already existing language version to a non existing language version, then you have the content of the old language and it is quite annoying to have to delete everything manually before you start typing. A tradeoff would be: When we go from a saved language version to a non existing language the content is deleted. Then if we add values to the text and title, the fields are not emptied until we click on an existing language version in the DB. What do you think ? This is what is currently hapening with the changes I made. But I am happy to update that if you think always keeping the previous content is better.

suzinyou replied:
I see.. I think from user's perspective a smoother experience might be

1. Add new language -- app changes state to _editing_ or something
2. Add title & text
3. Click on English (without clicking Save) -- app asks if you want to save changes. 
4. Click on Save
5. Then you are able to click on English

If this doesn't seem like a 10min job, let's create a backlog ticket.

suzinyou replied:
Aha. Relatedly, when I'm on Add Content page,

1. POST /language to create new language (e.g. HINDI)
1. Without filling out title or body, click on language ➕  and click Hindi
  - This removes the English tab (bc it was empty, I guess?) . 
3. Now enter some title and body in Hindi tab
4. Click language ➕ button and click English
5. **Now English tab is filled with same content as 3.**


I think this deserves a more thorough thinking around UI/UX. cc: @Tanmay-97

suzinyou replied:
Another reason why I think we need to thoroughly consider better UI/UX is because of the delete button's functionality under each language tab.

Right now, 
1. when it's an unsaved language version, delete button just deletes the language tab.
2. If it's a saved language version with no other language version, the same button can delete the entire content. 

We need some visual distinction.. For example, don't allow deleting last remaining  language version in an edit page. And/or have a separate content-level delete button outside the language tabs.

<img width="1113" alt="image" src="https://github.com/IDinsight/aaq-core/assets/7042047/23a343b0-4eeb-4c7b-a839-726cf4ceff46">

Also, it seems too powerful to be able to delete an entire content from just content card landing page?! But that's up for debate

lickem22 replied:
I think this is something thats going to be hard to discuss by text let's talk about it during our checkin. Then I can make the changes later

lickem22 commented:
> Thanks @lickem22 - we are almost there. Left a few comments.
> 
> Also, can you please fix these errors in the console: <img alt="image" width="651" src="https://private-user-images.githubusercontent.com/14125957/320269129-949b744d-ed78-43b5-b854-7da6e7ecb8f0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI1NzU3MDQsIm5iZiI6MTcxMjU3NTQwNCwicGF0aCI6Ii8xNDEyNTk1Ny8zMjAyNjkxMjktOTQ5Yjc0NGQtZWQ3OC00M2I1LWI4NTQtN2RhNmU3ZWNiOGYwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA4VDExMjMyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFkMTBjZGE2M2I5OGMzYzY0Yzg2OTAxNjhhMzg0YjQ0OGQwMWFmYjM3ZTBmMjg0NmZiYjkyNWU0N2YzMzQ1NGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.C408oX4h6bfKOkyDFpRCW-UmIOrvgGNBA0u5DY8Smco">
> 
> <img alt="image" width="901" src="https://private-user-images.githubusercontent.com/14125957/320269162-b8cab8c6-920a-4f2c-a53c-af678ae3999d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI1NzU3MDQsIm5iZiI6MTcxMjU3NTQwNCwicGF0aCI6Ii8xNDEyNTk1Ny8zMjAyNjkxNjItYjhjYWI4YzYtOTIwYS00ZjJjLWE1M2MtYWY2NzhhZTM5OTlkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA4VDExMjMyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI5Zjk0ODMwM2ZiZmMwMGMxZTQ0MjhjYzg4YWVkMzNkMGQ3ZGM3NmQyNmJlNWI1NTJmOTI0ZTdkZWEwYzQyNzcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.REIjHpz-rp9CfWStfx48W85KwHYnEzb9NfPg_5stMPo"> <img alt="image" width="884" src="https://private-user-images.githubusercontent.com/14125957/320269174-42ac23f9-9aa7-4b00-b144-906a61e89869.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI1NzU3MDQsIm5iZiI6MTcxMjU3NTQwNCwicGF0aCI6Ii8xNDEyNTk1Ny8zMjAyNjkxNzQtNDJhYzIzZjktOWFhNy00YjAwLWIxNDQtOTA2YTYxZTg5ODY5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA4VDExMjMyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI4NjM2ODI3N2IxMGFhOTY0YmEzYjlmOTJiZmQ0ZjEyOTQ2NzZkYzRkMGRjNTJmMWJhODlhMDEwMWUyZDc1MWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.oJSlSkdCbPtzh1Ez2vJRys9U8-Rcj2hyHiQRw5M7AYQ">

Thanks Sid. I fixed most of your requests and added a few comments. Also I fixed the first bug. Howeve r I was not able to recreate the two last errors, could you let me know how to reproduce them so I can fix?

suzinyou commented on admin_app/src/app/content/edit/page.tsx:
This is very cool @lickem22  :-)

Do you know if there's an easy way to disable `Save` button until after API response arrives?

Notice below I was able to add multiple Hindi versions of the same content ID!

<img width="858" alt="image" src="https://github.com/IDinsight/aaq-core/assets/7042047/b6febe35-6210-41ff-af4c-6ed56e5552d3">

Steps to replicate
1. POST /language ("HINDI")
2. Create some English content
3. For one content, Edit
4. Add Hindi language version
5. Fill in title and body. Hit Save -- nothing changes so I hit save multiple times
6. In the landing page, filter for Hindi -- Same content ID, multiple Hindi versions!

suzinyou replied:
This seems important to fix because now /delete, /edit etc. won't work.. because 2 rows are retrieved!

> sqlalchemy.exc.MultipleResultsFound: Multiple rows were found when one or none was required

You can only delete the whole content, not language-by-language..

lickem22 replied:
But that is a backend issue right ? It should not allow that. But I checked my code and it definitely looks out for that. Maybe it's an async issue. I will look into it further.

suzinyou replied:
Checked that it works!

suzinyou reviewed (changes requested):
Hello @lickem22 , thank you so much for this!! It's looking great.. though we should rethink language related UI/UX. Perhaps something we can get outside help on :-)

I am requesting just 1 change, other ones please address/create tickets as you see fit. Let me know what you think.

lickem22 commented on admin_app/package.json:
I really don't know what's happening here. The code seems pretty much the same to me. I'll double check

suzinyou replied:
I pushed

suzinyou reviewed (approved):
Thanks Carlos!
